File: sdks/kotlin/src/main/kotlin/com/neuratalk/sdk/NeuraTalkClient.kt
package com.neuratalk.sdk

import org.json.JSONArray
import org.json.JSONObject
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

private val RETRYABLE_STATUS = setOf(429, 500, 502, 503, 504)

/**
 * Official NeuraTalk API client. Auth endpoints (register/login/requestOtp/
 * verifyOtp) don't require a token; everything else does — pass one via the
 * constructor or [withToken].
 */
class NeuraTalkClient(
    private val token: String? = null,
    private val baseUrl: String = "https://neuratalk.in",
    private val maxRetries: Int = 3,
    private val retryBaseDelayMs: Long = 300,
    private val httpClient: HttpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(30)).build(),
) {
    fun withToken(token: String): NeuraTalkClient =
        NeuraTalkClient(token, baseUrl, maxRetries, retryBaseDelayMs, httpClient)

    // ---- Auth ----

    fun register(username: String, password: String, extra: JSONObject = JSONObject()): JSONObject {
        val body = JSONObject(extra.toString()).put("username", username).put("password", password)
        return request("POST", "/api/auth/register", body = body, auth = false)
    }

    fun login(username: String, password: String): JSONObject =
        request("POST", "/api/auth/login", body = JSONObject().put("username", username).put("password", password), auth = false)

    fun requestOtp(identifier: String, channel: String): JSONObject =
        request("POST", "/api/auth/otp/request", body = JSONObject().put("identifier", identifier).put("channel", channel), auth = false)

    fun verifyOtp(identifier: String, channel: String, code: String, firebaseToken: String? = null): JSONObject {
        val body = JSONObject().put("identifier", identifier).put("channel", channel).put("code", code)
        if (firebaseToken != null) body.put("firebaseToken", firebaseToken)
        return request("POST", "/api/auth/otp/verify", body = body, auth = false)
    }

    fun me(): JSONObject = request("GET", "/api/auth/me")

    fun logout() {
        request("POST", "/api/auth/logout")
    }

    // ---- Calls ----

    fun createCall(
        calleeIdentifier: String,
        callType: String,
        myLanguage: String = "auto",
        theirLanguage: String = "auto",
    ): JSONObject {
        val body = JSONObject()
            .put("calleeIdentifier", calleeIdentifier)
            .put("callType", callType)
            .put("myLanguage", myLanguage)
            .put("theirLanguage", theirLanguage)
        return request("POST", "/api/calls/create", body = body)
    }

    fun createConferenceCall(participantIds: List<String>, hostLanguage: String = "auto", title: String? = null): JSONObject {
        val body = JSONObject().put("participantIds", JSONArray(participantIds)).put("hostLanguage", hostLanguage)
        if (title != null) body.put("title", title)
        return request("POST", "/api/calls/conference", body = body)
    }

    fun connectCall(callId: String, receiverNumber: String? = null): JSONObject {
        val body = JSONObject()
        if (receiverNumber != null) body.put("receiverNumber", receiverNumber)
        return request("POST", "/api/calls/${enc(callId)}/connect", body = body)
    }

    fun endCall(callId: String): JSONObject = request("POST", "/api/calls/${enc(callId)}/end")

    fun holdCall(callId: String): JSONObject = request("POST", "/api/calls/${enc(callId)}/hold")

    fun resumeCall(callId: String): JSONObject = request("DELETE", "/api/calls/${enc(callId)}/hold")

    fun rejectCall(callId: String): JSONObject = request("POST", "/api/calls/${enc(callId)}/reject")

    fun getIncomingCall(): JSONObject? {
        val res = request("GET", "/api/calls/incoming")
        return if (res.isNull("incoming") || !res.has("incoming")) null else res.optJSONObject("incoming")
    }

    fun listCallHistory(limit: Int = 50): JSONArray =
        request("GET", "/api/calls/history?limit=$limit").optJSONArray("calls") ?: JSONArray()

    fun getCall(callId: String): JSONObject = request("GET", "/api/calls/${enc(callId)}")

    // ---- Transcripts ----

    fun searchTranscripts(query: String, limit: Int? = null, offset: Int? = null): JSONObject {
        val params = StringBuilder("q=${enc(query)}")
        if (limit != null) params.append("&limit=$limit")
        if (offset != null) params.append("&offset=$offset")
        return request("GET", "/api/transcripts/search?$params")
    }

    fun getTranscript(callId: String): JSONArray =
        request("GET", "/api/transcripts/${enc(callId)}").optJSONArray("segments") ?: JSONArray()

    fun deleteTranscript(callId: String): JSONObject = request("DELETE", "/api/transcripts/${enc(callId)}")

    /** Returns the raw exported file bytes — write them to disk or stream them as-is. */
    fun exportTranscript(callId: String, format: String): ByteArray {
        val res = rawRequest("GET", "/api/transcripts/${enc(callId)}/export/$format", bodyBytes = null)
        return res.body()
    }

    // ---- Internals ----

    private fun enc(s: String): String = URLEncoder.encode(s, StandardCharsets.UTF_8)

    private fun request(method: String, path: String, body: JSONObject? = null, auth: Boolean = true): JSONObject {
        val bodyBytes = body?.toString()?.toByteArray(StandardCharsets.UTF_8)
        val res = rawRequest(method, path, bodyBytes, auth)
        val text = String(res.body(), StandardCharsets.UTF_8)
        return if (text.isBlank()) JSONObject() else JSONObject(text)
    }

    private fun rawRequest(method: String, path: String, bodyBytes: ByteArray?, auth: Boolean = true): HttpResponse<ByteArray> {
        val url = URI.create("$baseUrl$path")
        var attempt = 0
        while (true) {
            attempt++
            val builder = HttpRequest.newBuilder(url).timeout(Duration.ofSeconds(30))
            if (bodyBytes != null) builder.header("Content-Type", "application/json")
            if (auth && token != null) builder.header("Authorization", "Bearer $token")
            val publisher = if (bodyBytes != null) HttpRequest.BodyPublishers.ofByteArray(bodyBytes) else HttpRequest.BodyPublishers.noBody()
            builder.method(method, publisher)

            val res: HttpResponse<ByteArray>
            try {
                res = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofByteArray())
            } catch (cause: Exception) {
                if (attempt > maxRetries) throw NeuraTalkNetworkException(cause)
                backoff(attempt)
                continue
            }

            if (res.statusCode() !in 200..299) {
                if (res.statusCode() in RETRYABLE_STATUS && attempt <= maxRetries) {
                    backoff(attempt, res.headers().firstValue("Retry-After").orElse(null))
                    continue
                }
                val errorBody = try {
                    val text = String(res.body(), StandardCharsets.UTF_8)
                    if (text.isBlank()) null else JSONObject(text)
                } catch (_: Exception) {
                    null
                }
                throw NeuraTalkApiException(res.statusCode(), errorBody)
            }
            return res
        }
    }

    private fun backoff(attempt: Int, retryAfterHeader: String? = null) {
        val delayMs = retryAfterHeader?.toDoubleOrNull()?.let { (it * 1000).toLong() }
            ?: (retryBaseDelayMs * (1L shl (attempt - 1)))
        Thread.sleep(delayMs)
    }
}

File: sdks/swift/Sources/NeuraTalkSDK/NeuraTalkClient.swift
import Foundation

private let retryableStatus: Set<Int> = [429, 500, 502, 503, 504]

/// Official NeuraTalk API client. Auth endpoints (`register`/`login`/
/// `requestOtp`/`verifyOtp`) don't require a token; everything else does —
/// pass one via the initializer or `withToken(_:)`.
public final class NeuraTalkClient {
    private let token: String?
    private let baseUrl: String
    private let maxRetries: Int
    private let retryBaseDelay: TimeInterval
    private let session: URLSession

    public init(
        token: String? = nil,
        baseUrl: String = "https://neuratalk.in",
        maxRetries: Int = 3,
        retryBaseDelay: TimeInterval = 0.3,
        session: URLSession = .shared
    ) {
        self.token = token
        self.baseUrl = baseUrl.hasSuffix("/") ? String(baseUrl.dropLast()) : baseUrl
        self.maxRetries = maxRetries
        self.retryBaseDelay = retryBaseDelay
        self.session = session
    }

    /// Returns a new client bound to the given session token, leaving this one unmodified.
    public func withToken(_ token: String) -> NeuraTalkClient {
        NeuraTalkClient(token: token, baseUrl: baseUrl, maxRetries: maxRetries, retryBaseDelay: retryBaseDelay, session: session)
    }

    // MARK: - Auth

    public func register(username: String, password: String, extra: [String: Any] = [:]) async throws -> [String: Any] {
        var body = extra
        body["username"] = username
        body["password"] = password
        return try await request("POST", "/api/auth/register", body: body, auth: false)
    }

    public func login(username: String, password: String) async throws -> [String: Any] {
        try await request("POST", "/api/auth/login", body: ["username": username, "password": password], auth: false)
    }

    public func requestOtp(identifier: String, channel: String) async throws -> [String: Any] {
        try await request("POST", "/api/auth/otp/request", body: ["identifier": identifier, "channel": channel], auth: false)
    }

    public func verifyOtp(identifier: String, channel: String, code: String, firebaseToken: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = ["identifier": identifier, "channel": channel, "code": code]
        if let firebaseToken { body["firebaseToken"] = firebaseToken }
        return try await request("POST", "/api/auth/otp/verify", body: body, auth: false)
    }

    public func me() async throws -> [String: Any] {
        try await request("GET", "/api/auth/me")
    }

    public func logout() async throws {
        _ = try await request("POST", "/api/auth/logout")
    }

    // MARK: - Calls

    public func createCall(
        calleeIdentifier: String,
        callType: String,
        myLanguage: String = "auto",
        theirLanguage: String = "auto"
    ) async throws -> [String: Any] {
        try await request("POST", "/api/calls/create", body: [
            "calleeIdentifier": calleeIdentifier,
            "callType": callType,
            "myLanguage": myLanguage,
            "theirLanguage": theirLanguage,
        ])
    }

    public func createConferenceCall(participantIds: [String], hostLanguage: String = "auto", title: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = ["participantIds": participantIds, "hostLanguage": hostLanguage]
        if let title { body["title"] = title }
        return try await request("POST", "/api/calls/conference", body: body)
    }

    public func connectCall(callId: String, receiverNumber: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = [:]
        if let receiverNumber { body["receiverNumber"] = receiverNumber }
        return try await request("POST", "/api/calls/\(encode(callId))/connect", body: body)
    }

    public func endCall(callId: String) async throws -> [String: Any] {
        try await request("POST", "/api/calls/\(encode(callId))/end")
    }

    public func holdCall(callId: String) async throws -> [String: Any] {
        try await request("POST", "/api/calls/\(encode(callId))/hold")
    }

    public func resumeCall(callId: String) async throws -> [String: Any] {
        try await request("DELETE", "/api/calls/\(encode(callId))/hold")
    }

    public func rejectCall(callId: String) async throws -> [String: Any] {
        try await request("POST", "/api/calls/\(encode(callId))/reject")
    }

    public func getIncomingCall() async throws -> [String: Any]? {
        let res = try await request("GET", "/api/calls/incoming")
        return res["incoming"] as? [String: Any]
    }

    public func listCallHistory(limit: Int = 50) async throws -> [[String: Any]] {
        let res = try await request("GET", "/api/calls/history?limit=\(limit)")
        return res["calls"] as? [[String: Any]] ?? []
    }

    public func getCall(callId: String) async throws -> [String: Any] {
        try await request("GET", "/api/calls/\(encode(callId))")
    }

    // MARK: - Transcripts

    public func searchTranscripts(query: String, limit: Int? = nil, offset: Int? = nil) async throws -> [String: Any] {
        var items = [URLQueryItem(name: "q", value: query)]
        if let limit { items.append(URLQueryItem(name: "limit", value: String(limit))) }
        if let offset { items.append(URLQueryItem(name: "offset", value: String(offset))) }
        var components = URLComponents()
        components.queryItems = items
        // URLComponents percent-encodes "+" as a literal plus, which some
        // servers misread as a space in query strings — encode it explicitly.
        let query = (components.percentEncodedQuery ?? "").replacingOccurrences(of: "+", with: "%2B")
        return try await request("GET", "/api/transcripts/search?\(query)")
    }

    public func getTranscript(callId: String) async throws -> [[String: Any]] {
        let res = try await request("GET", "/api/transcripts/\(encode(callId))")
        return res["segments"] as? [[String: Any]] ?? []
    }

    public func deleteTranscript(callId: String) async throws -> [String: Any] {
        try await request("DELETE", "/api/transcripts/\(encode(callId))")
    }

    /// Returns the raw exported file bytes — write them to disk or stream them as-is.
    public func exportTranscript(callId: String, format: String) async throws -> Data {
        try await rawRequest("GET", "/api/transcripts/\(encode(callId))/export/\(format)").0
    }

    // MARK: - Internals

    private func encode(_ s: String) -> String {
        s.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? s
    }

    private func request(_ method: String, _ path: String, body: [String: Any]? = nil, auth: Bool = true) async throws -> [String: Any] {
        let (data, _) = try await rawRequest(method, path, body: body, auth: auth)
        guard !data.isEmpty else { return [:] }
        let parsed = try? JSONSerialization.jsonObject(with: data)
        return parsed as? [String: Any] ?? [:]
    }

    private func rawRequest(_ method: String, _ path: String, body: [String: Any]? = nil, auth: Bool = true) async throws -> (Data, HTTPURLResponse) {
        guard let url = URL(string: "\(baseUrl)\(path)") else {
            throw NeuraTalkApiError(status: 0, body: ["error": "Invalid URL: \(path)"])
        }

        var attempt = 0
        while true {
            attempt += 1
            var req = URLRequest(url: url)
            req.httpMethod = method
            if auth, let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
            if let body {
                req.setValue("application/json", forHTTPHeaderField: "Content-Type")
                req.httpBody = try? JSONSerialization.data(withJSONObject: body)
            }

            let data: Data
            let response: URLResponse
            do {
                (data, response) = try await session.data(for: req)
            } catch {
                if attempt > maxRetries { throw NeuraTalkNetworkError(cause: error) }
                await backoff(attempt)
                continue
            }

            guard let http = response as? HTTPURLResponse else {
                throw NeuraTalkNetworkError(cause: URLError(.badServerResponse))
            }

            if !(200..<300).contains(http.statusCode) {
                if retryableStatus.contains(http.statusCode) && attempt <= maxRetries {
                    let retryAfter = http.value(forHTTPHeaderField: "Retry-After")
                    await backoff(attempt, retryAfterHeader: retryAfter)
                    continue
                }
                let errorBody = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
                throw NeuraTalkApiError(status: http.statusCode, body: errorBody)
            }
            return (data, http)
        }
    }

    private func backoff(_ attempt: Int, retryAfterHeader: String? = nil) async {
        let delay: TimeInterval
        if let retryAfterHeader, let seconds = Double(retryAfterHeader) {
            delay = seconds
        } else {
            delay = retryBaseDelay * pow(2, Double(attempt - 1))
        }
        try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
    }
}
